Add CTest support for greentea tests: events

hazzlim opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Fixes #14965 

- the `MBED_EXTENDED_TESTS` macro is removed and 'extended' tests are enabled by default
- Multi-part FAT filesystem tests are refactored, splitting the test case with large ROM requirements into a separate suite in order to isolate it for skipping.
- `events/equeue` tests that require RTOS are wrapped in checks for RTOS
- events greentea tests are refactored to use std::chrono values and remove deprecated APIs
- CMake support for events greentea tests is refactored, following pilot work, to build as part of global project when `BUILD_GREENTEA_TESTS` is set. Added checks for macros that indicate the test is not supported and conditionally set the `TEST_SKIPPED` argument accordingly.

Preceding:
- #14892 
- #14902

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None
### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@ARMmbed/mbed-os-core 
----------------------------------------------------------------------------------------------------------------

hazzlim committed:
CMake: Greentea: Port events tests to CTest

Following the approach defined in #14892 and #14902, events greentea
tests are ported to enable running with CTest.

events/queue and events/timing both require DEVICE_USTICKER.

LDong-Arm commented on events/tests/TESTS/events/equeue/CMakeLists.txt:
```diff
@@ -1,20 +1,19 @@
-include(${MBED_PATH}/tools/cmake/mbed_greentea.cmake)
-
-project(${TEST_TARGET})
+if(NOT "MBED_EXTENDED_TESTS" IN_LIST MBED_CONFIG_DEFINITIONS)
```
If I remember correctly, the `MBED_EXTENDED_TESTS` macro was created to disable tests that are slow to run by default, to save CI time. But if a user want to enable all tests, they need to define it. How will they be able to do it?

Mbed CLI 1 has a `-D<macro>` option for this, and we don't have an equivalent in Mbed CLI 2. I can think of three options:
* add a `-D<macro>` option to mbed-tools, and generated mbed_config.cmake will contain add `<macro>` to `MBED_CONFIG_DEFINITIONS` which eventual get [passed to `target_compile_definitions`](https://github.com/ARMmbed/mbed-os/blob/68667df7ac76a2c2660b331a350d1a2ceaa433b7/CMakeLists.txt#L102-L106) as _C/C++ macros_
* add a `-D<macro>` option to mbed-tools, but `<macro>` becomes a CMake variable (rather than a C/C++ macro). Each test's CMakeLists.txt can check this variable
* add `TESTS/configs/extended.json` containing
    ```json
    {
        "macros": ["MBED_EXTENDED_TESTS"]
    }
    ```
    so users can pass `--app-config TESTS/configs/extended.json` to `mbed-tools configure`.

What do you and @ARMmbed/mbed-os-core think?

LDong-Arm replied:
General question on mbed-tools in general, not just tests: Do we want to give users ability to set a global C/C++ macro on the fly, like Mbed CLI 1 does?

Patater replied:
The concept of `MBED_EXTENDED_TESTS` seems flawed to me. If we want to save time, we should do it by either making tests run quicker, or by increasing our reliance on unit tests and de-emphasizing integration tests, not by avoiding running tests.

How about we always include `MBED_EXTENDED_TESTS` type tests when using CLI 2? This would mean no mention of `MBED_EXTENDED_TESTS` in CMake.

rwalton-arm replied:
> General question on mbed-tools in general, not just tests: Do we want to give users ability to set a global C/C++ macro on the fly, like Mbed CLI 1 does?

I think we should just let users run CMake directly instead of adding some pass-through for global definitions. mbed-tools is supposed to be "easy mode", if you want "advanced mode" you can use cmake directly.

Patater replied:
We may also need to remove `MBED_EXTENDED_TESTS` from C or C++ code too, to make the tests work with CMake. This has a side effect of always running the tests for CLI 1 as well, but that's acceptable.

hazzlim replied:
Is there a cost consideration when it comes to the time it takes to run integration tests in the CI? I was under the impression that they run overnight, so time itself wouldn't be a major concern on its own.

rwalton-arm replied:
> The concept of `MBED_EXTENDED_TESTS` seems flawed to me. If we want to save time, we should do it by either making tests run quicker, or by increasing our reliance on unit tests and de-emphasizing integration tests, not by avoiding running tests.
> 
> How about we always include `MBED_EXTENDED_TESTS` type tests when using CLI 2? This would mean no mention of `MBED_EXTENDED_TESTS` in CMake.

This seems reasonable to me. If the tests really take "too long" we should fix them so they run in a reasonable time.

LDong-Arm replied:
It might be interesting to try (e.g. on a local board) how long extended tests take, starting with events tests in this PR. We rarely use `MBED_EXTENDED_TESTS` except in a few places, so unless it really takes crazy long (which seems unlikely to me), this seems just to be inconsistency.

> The concept of MBED_EXTENDED_TESTS seems flawed to me. If we want to save time, we should do it by either making tests run quicker, or by increasing our reliance on unit tests and de-emphasizing integration tests, not by avoiding running tests.

+1

Patater replied:
The time saved in finding bugs or issues with a PR before it gets merged is well worth the extra time a PR job takes to run. When we are looking to make a release, we don't want to spend days tracking down why a test we never ran until the day before the scheduled release started failing: we want the failures up front and as close as possible to when the change is made, so it is easiest to fix (all context for the change is available and the person who raised the PR is still around to fix their PR rather than moved on to other things).

hazzlim replied:
From elapsed times reported by mbedgt:
- `events/equeue: 29.33 sec` (currently skipped w/o MBED_EXTENDED_TESTS defined)
- `fat_filesystem: 21.25 sec` (currently skipped w/o MBED_EXTENDED_TESTS defined)
- `multipart_fat_filesystem: 23.16 sec` (currently skipped w/o MBED_EXTENDED_TESTS defined)

For comparison: 
- `events/queue: 29.42` (not skipped
- `events/timing: 80.12 sec` (not skipped)

Do we think completely removing `MBED_EXTENDED_TESTS` from the codebase is the best way forward?

Patater replied:
> Do we think completely removing MBED_EXTENDED_TESTS from the codebase is the best way forward?

Yep. A few minutes of testing time is well worth it. Thanks for running the numbers so we can quantify this more easily.

hazzlim replied:
Rebased with `MBED_EXTENDED_TESTS` removal commit.

hazzlim committed:
Greentea: remove MBED_EXTENDED_TESTS macro

MBED_EXTENDED_TESTS macro was previously required to enable certain
Greentea tests, which were disabled by default in order to save CI time.

We remove this macro because the time saved in finding bugs or issues
with a PR prior to merging is worth the increased testing time.

hazzlim committed:
Greentea: Add macro check for RTOS in events/equeue

Some events/equeue Greentea test cases use the Thread API and therefore
require RTOS to run, but previously this was not checked in the code.
We add guards with the macro MBED_CONF_RTOS_PRESENT.

hazzlim committed:
Greentea: events: Remove deprecated APIs and use chrono

References to time should do so using std::chrono. We reworked events
tests to use std::chrono and new APIs in order to remove deprecation
warnings resulting from deprecated API calls. This required addition of
a macro for test assertions using std::chrono values.

Calls to wait_us() (not deprecated) are replaced with
ThisThread::sleep_for() as there is no reason to wait without sleeping.

Certain API calls that involve UserAllocatedEvents and do not have
std::chrono versions are left unchanged, for example
EventQueue.time_left().

LDong-Arm commented on events/tests/TESTS/events/equeue/CMakeLists.txt:
```diff
@@ -1,18 +1,11 @@
-# Copyright (c) 2020 ARM Limited. All rights reserved.
+# Copyright (c) 2021 ARM Limited. All rights reserved.
```
Not a huge problem, but we should extend the time rather than change it, i.e. 2020-2021

hazzlim replied:
Thanks Lingkai, I actually meant to ask this - is it preferred to change the copyright date (and extend as you advised) or to leave it as is?

LDong-Arm replied:
Even 2020 is not accurate, because events has been there for many years since early days of Mbed. We added a copyright retrospectively last year, so the start date is okay I guess..

LDong-Arm reviewed (commented):
CI failed because `test_equeue_multithread` and `test_equeue_multithreaded_barrage` use `Thread` which needs RTOS and doesn't work with bare metal. You can guard them with `#ifdef MBED_CONF_RTOS_PRESENT`.

Note: `Thread` requires RTOS, but `ThisThread` works fine with bare metal because bare metal is treated like only one thread (i.e. `ThisThread`).

It's nice that our CMake test includes a bare metal target.

Patater commented on events/tests/TESTS/events/timing/main.cpp:
```diff
@@ -128,7 +152,8 @@ void semaphore_timing_test()
 utest::v1::status_t test_setup(const size_t number_of_cases)
 {
-    GREENTEA_SETUP((number_of_cases + 1)*TEST_EVENTS_TIMING_TIME / 1000, "default_auto");
+    GREENTEA_SETUP((number_of_cases + 1) * duration_cast<seconds>(TEST_EVENTS_TIMING_TIME_DURATION).count(),
```
Looks like we had some bad style before. Is `events/tests/TESTS/events/timing` somehow exclused from `astyle` checking?

hazzlim replied:
It’s not excluded...but I just tried running astyle on it (the master branch version) and it returned “Unchanged”.  Maybe `.astylerc` needs updating?

hazzlim committed:
Greentea: Separate large FAT test

The Greentea test suite multipart_fat_filesystem contains a test case
that uses LittleFileSystem and FATFilesystem, and including this case
results in a binary too large for the flash memory of some targets.

We separate this test case into its own suite and guard this test suite
with a check for the size of the target's ROM. This is chosen over
simply guarding the individual test case inside the same suite as it
will allow the newly created suite to be skipped from CMake when the
FAT Filesystem tests are ported to CTest.

This issue was previously unidentified because the test suite was
disabled by default, requiring the `MBED_EXTENDED_TESTS` macro to be
defined, and was therefore not run in CI.

hazzlim committed:
Greentea: Refactor FAT Filesystem test cases to be independent

In the Greentea test suite `multipart_fat_filesystem`, all the test
cases are dependent on setup that occurs in the first case. This case
formats (with FAT) a HeapBlockDevice that is shared between cases.
Alongside this, the first case checks that sufficient heap memory is
available for the dynamic memory allocation required for the FAT format
process. This check is performed by testing if we can successfully
allocate the required number of bytes, and skipping the case if not. The
check is then propagated to the other cases by skipping those if the
HeapBlockDevice pointer is falsy, which it will be if the first test is
skipped.

We refactor the cases to be independent of one another. Now each test
case checks that sufficient memory can be allocated for FAT formatting,
performing said formatting or skipping as indicated by this check.

In doing so we replace remove some raw pointers, and replace some C
style arrays with std::array, in order to make memory management less
error prone.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -18,6 +18,7 @@
 #include "greentea-client/test_env.h"
 #include "unity.h"
 #include "utest.h"
+#include <memory>
```
In commit message "Greentea: Refactor FAT Filesystem test cases to be independent" there is a typo: "propogated" should be "propagated"

hazzlim replied:
Fixed.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -32,36 +33,57 @@ using namespace utest::v1;
+bool check_heap_allocation(size_t bytes)
+{
+    unique_ptr<uint8_t[]> dummy {new (std::nothrow) uint8_t[bytes]};
+    return (dummy != nullptr);
```
We should just be able to `return dummy.get()` here and a `nullptr` will be deduced as `false`

hazzlim replied:
Good point! I have amended with this change.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -32,36 +33,57 @@ using namespace utest::v1;
+void format_partitions(HeapBlockDevice &bd)
+{
+    // Create two partitions splitting device in ~half
+    TEST_ASSERT_EQUAL(0, MBRBlockDevice::partition(&bd, 1, 0x83, 0, (BLOCK_COUNT / 2) * BLOCK_SIZE));
```
`0x83` should be a named variable.

hazzlim replied:
Done.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -85,14 +107,18 @@ void test_format()
-    TEST_SKIP_UNLESS_MESSAGE(bd, "Not enough heap memory to run test. Test skipped.");
+    TEST_SKIP_UNLESS_MESSAGE(check_heap_allocation(mem_alloc_threshold), "Not enough heap memory to run test. Test skipped.");
+
+    HeapBlockDevice bd {BLOCK_COUNT * BLOCK_SIZE, BLOCK_SIZE};
```
This test case seems to make lots of assertions. Any way we can split it up into separate cases?

hazzlim replied:
I agree it's a pretty huge number of assertions. I'm not really sure how to split it up however, as verifying the data read matches the data written requires all the steps in the case as far as I can see.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -248,10 +267,10 @@ void test_with_other_fs()
             TEST_ASSERT_EQUAL(0, err);
 
-            part = new (std::nothrow) MBRBlockDevice(bd, 1);
+            part = new (std::nothrow) MBRBlockDevice(&bd, 1);
```
Can we use a smart pointer instead of raw `new` and `delete`?

rwalton-arm replied:
This was removed in a later commit.

hazzlim replied:
Yep, I've removed the raw pointers in the commit "Greentea: Refactor test with other filesystem".

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem/main.cpp:
```diff
@@ -23,7 +23,6 @@
 #include "HeapBlockDevice.h"
```
Commit message "Greentea: Seperate large FAT test" contains a typo: "seperate" should be "separate".

hazzlim replied:
Fixed.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem_with_other_filesystem/main.cpp:
```diff
@@ -0,0 +1,142 @@
+bool check_heap_allocation(size_t bytes)
+{
+    unique_ptr<uint8_t[]> dummy {new (std::nothrow) uint8_t[bytes]};
+    return (dummy != nullptr);
```
We should be able to just `return dummy.get()` here.

hazzlim replied:
Good point! I have amended with this change.

rwalton-arm commented on storage/filesystem/fat/tests/TESTS/fat/multipart_fat_filesystem_with_other_filesystem/main.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    TEST_SKIP_UNLESS_MESSAGE(check_heap_allocation(mem_alloc_threshold), "Not enough heap memory to run test. Test skipped.");
+
+    // Stage 0 - LittleFS
```
Could these "Stages" be separate test cases? The loop looks pretty awkward at the moment with all this "if stage==1" stuff.

hazzlim replied:
Yes it's not very readable! I have split the stages into three separate cases in the commit "Greentea: Refactor test with other filesystem".

rwalton-arm commented on events/tests/TESTS/events/timing/main.cpp:
```diff
@@ -57,6 +62,16 @@ float chisq(float sigma)
 }
 
+// Macro for random duration generation
+#define RANDOM_DURATION(sigma) \
```
Could this be a normal function instead of a macro?

hazzlim replied:
Yes good point, I have replaced the macro with a function.

rwalton-arm commented on events/tests/TESTS/events/timing/main.cpp:
```diff
@@ -83,14 +101,18 @@ void timer_timing_test()
-    unsigned start = equeue_tick();
-    int prev = 0;
+    // equeue_tick() returns an integral millisecond value
+    auto start = milliseconds{equeue_tick()};
```
Looks like `start` should be `const`?

hazzlim replied:
Done.

rwalton-arm commented on events/tests/TESTS/events/timing/main.cpp:
```diff
@@ -83,14 +101,18 @@ void timer_timing_test()
-    while (prev < TEST_EVENTS_TIMING_TIME) {
-        int next = equeue_tick() - start;
+    while (prev < TEST_EVENTS_TIMING_TIME_DURATION) {
+        auto next = milliseconds{equeue_tick()} - start;
```
Looks like `next` should be `const`?

hazzlim replied:
Done.

rwalton-arm commented on events/tests/TESTS/events/timing/main.cpp:
```diff
@@ -107,18 +129,20 @@ void semaphore_timing_test()
-        int taken = timer.read_us() - start;
+        auto start = timer.elapsed_time();
+        equeue_sema_wait(&sema, duration_cast<milliseconds>(delay).count());
+        auto taken = timer.elapsed_time() - start;
```
Looks like most, if not all, of these variables should be `const`?

hazzlim replied:
Done.

hazzlim committed:
Greentea: Refactor test with other filesystem

The test case `test_with_other_fs` uses a fairly awkward loop to test
mounting different filesystems consecutively.

We split the test case into seperate cases, using more descriptive names
and removing the use of raw pointers.
